src/ln/msgs.rs

```diff
@@ -382,9 +384,93 @@ impl NetAddress {
 			&NetAddress::OnionV3 {..} => { 4 },
 		}
 	}
+
+	fn len(&self) -> u16 {
+		match self {
+			&NetAddress::IPv4 { .. } => { 6 },
+			&NetAddress::IPv6 { .. } => { 18 },
+			&NetAddress::OnionV2 { .. } => { 12 },
+			&NetAddress::OnionV3 { .. } => { 37 },
+		}
+	}
 }
 
-#[derive(Clone)]
+impl Writeable for NetAddress {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		match self {
+			&NetAddress::IPv4 { ref addr, ref port } => {
+				0u8.write(writer)?;
+				addr.write(writer)?;
+				port.write(writer)?;
+			},
+			&NetAddress::IPv6 { ref addr, ref port } => {
+				1u8.write(writer)?;
+				addr.write(writer)?;
+				port.write(writer)?;
+			},
+			&NetAddress::OnionV2 { ref addr, ref port } => {
+				2u8.write(writer)?;
+				addr.write(writer)?;
+				port.write(writer)?;
+			},
+			&NetAddress::OnionV3 { ref ed25519_pubkey, ref checksum, ref version, ref port } => {
+				3u8.write(writer)?;
+				ed25519_pubkey.write(writer)?;
+				checksum.write(writer)?;
+				version.write(writer)?;
+				port.write(writer)?;
+			}
+		}
+		Ok(())
+	}
+}
+
+impl<R: ::std::io::Read>  Readable<R> for NetAddress {
+	fn read(reader: &mut R) -> Result<NetAddress, DecodeError> {
+		let byte = <u8 as Readable<R>>::read(reader)?;
+		match byte {
+			0 => {
+				let addr = Readable::read(reader)?;
+				let port = Readable::read(reader)?;
+				Ok(NetAddress::IPv4 {
+					addr,
+					port,
+				})
+			},
+			1 => {
+				let addr = Readable::read(reader)?;
+				let port = Readable::read(reader)?;
+				Ok(NetAddress::IPv6 {
+					addr,
+					port,
+				})
+			},
+			2 => {
+				let addr = Readable::read(reader)?;
+				let port = Readable::read(reader)?;
+				Ok(NetAddress::OnionV2 {
+					addr,
+					port,
+				})
+			},
+			3 => {
+				let ed25519_pubkey = Readable::read(reader)?;
+				let checksum = Readable::read(reader)?;
+				let version = Readable::read(reader)?;
+				let port = Readable::read(reader)?;
+				Ok(NetAddress::OnionV3 {
+					ed25519_pubkey,
+					checksum,
+					version,
+					port,
+				})
+			}
+			_ => return Err(DecodeError::UnknownVersion { byte }),
```

Comment:
Doesn't InvalidValue make more sense here?

Reply:
To me InvalidValue is data that is syntaxically incorrect at the protocol level, there address descriptor is may be right we just don't support it.

Reply:
We use InvalidValue mostly to track non-0/1 for bool values, which seems kinda equivalent here - the type byte isnt a "version" so much as something analogous to a 2 for a bool